File: 101. Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root or not root.left and not root.right:
            return True
        if not root.left or not root.right:
            return False
        return self.helper(root.left, root.right)
        
    def helper(self, p, q):
        if not p and not q:
            return True
        if not p or not q:
            return False
        return p.val == q.val and self.helper(p.left, q.right) and self.helper(q.left, p.right)

File: 101. Symmetric Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    /**
     * recursive method
     * @param root
     * @return
     */
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return isSym(root.left, root.right);
    }
    private boolean isSym(TreeNode l, TreeNode r) {
        if (l==null && r==null) {
            return true;
        } else if (l==null || r==null) {
            return false;
        } else {
            return isSym(l.left, r.right) && isSym(l.right, r.left) && l.val==r.val;
        }
    }

    /**
     * iterative method, notice ArrayDeque in java don't allow adding null as element, use LinkedList instead
     * @param root
     * @return
     */
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        Deque<TreeNode> q = new LinkedList<>();
        q.add(root.left);
        q.add(root.right);
        while (!q.isEmpty()) {
            TreeNode l = q.poll();
            TreeNode r = q.poll();
            if (l==null && r==null) {
                continue;
            } else if (l==null || r==null) {
                return false;
            } else if (l.val != r.val) {
                return false;
            } else {
                q.add(l.left);
                q.add(r.right);
                q.add(l.right);
                q.add(r.left);
            }
        }
        return true;
    }
}